path_login.go

```diff
@@ -170,30 +175,69 @@ func (b *azureAuthBackend) verifyClaims(claims *additionalClaims, role *azureRol
 	return nil
 }
 
-func (b *azureAuthBackend) verifyResource(ctx context.Context, subscriptionID, resourceGroupName, vmName string, claims *additionalClaims, role *azureRole) error {
+func (b *azureAuthBackend) verifyResource(ctx context.Context, subscriptionID, resourceGroupName, vmName string, vmssName string, claims *additionalClaims, role *azureRole) error {
 	// If not checking anything with the resource id, exit early
-	if len(role.BoundResourceGroups) == 0 && len(role.BoundSubscriptionsIDs) == 0 && len(role.BoundLocations) == 0 {
+	if len(role.BoundResourceGroups) == 0 && len(role.BoundSubscriptionsIDs) == 0 && len(role.BoundLocations) == 0 && len(role.BoundScaleSets) == 0 {
 		return nil
 	}
 
-	if subscriptionID == "" || resourceGroupName == "" || vmName == "" {
-		return errors.New("subscription_id, resource_group_name, and vm_name are required")
+	if subscriptionID == "" || resourceGroupName == "" {
+		return errors.New("subscription_id and resource_group_name are required")
 	}
 
-	client := b.provider.ComputeClient(subscriptionID)
-	vm, err := client.Get(ctx, resourceGroupName, vmName, compute.InstanceView)
-	if err != nil {
-		return errwrap.Wrapf("unable to retrieve virtual machine metadata: {{err}}", err)
+	var principalID, location *string
+
+	switch {
+	case vmssName != "":
+		client := b.provider.VMSSClient(subscriptionID)
+		vmss, err := client.Get(ctx, resourceGroupName, vmssName)
+		if err != nil {
+			return errwrap.Wrapf("unable to retries virtual machine scale set metadata: {{err}}", err)
+		}
+
+		if vmss.Identity == nil {
+			return errors.New("vmss client did not return identity information")
+		}
+		if vmss.Identity.PrincipalID == nil {
+			return errors.New("vmss principal id is empty")
+		}
+
+		// Check bound scale sets
+		if len(role.BoundScaleSets) > 0 && !strListContains(role.BoundScaleSets, vmssName) {
+			return errors.New("scale set not authorized")
+		}
+
+		principalID = vmss.Identity.PrincipalID
+		location = vmss.Location
+
+	case vmName != "":
+		client := b.provider.ComputeClient(subscriptionID)
+		vm, err := client.Get(ctx, resourceGroupName, vmName, compute.InstanceView)
+		if err != nil {
+			return errwrap.Wrapf("unable to retrieve virtual machine metadata: {{err}}", err)
+		}
+
+		if vm.Identity == nil {
+			return errors.New("vm client did not return identity information")
+		}
+
+		if vm.Identity.PrincipalID == nil {
+			return errors.New("vm principal id is empty")
+		}
+
+		// Check bound scale sets
+		if len(role.BoundScaleSets) > 0 {
+			return errors.New("bound scale set defined but this vm isn't in a scale set")
+		}
+
+		principalID = vm.Identity.PrincipalID
+		location = vm.Location
+	default:
+		return errors.New("either vm_name or vmss_name is required")
```

Comment:
@chrishoffman Do you think providing _both_ names should also be considered an error?

Reply:
Agreed.  I had that comment at one point but it appears github ate it.

Reply:
@kalafut @chrishoffman thanks for the feedback. I've just added a check for that.